kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/commands/application/ApplicationCommandRegistry.kt

```diff
@@ -72,324 +72,114 @@ public open class ApplicationCommandRegistry : KoinComponent {
-        // Next, we need to associate all the commands we just registered with the commands in our extensions
-        toCreate.forEach { command ->
-            val match = response.first { command.matches(locale, it) }
-
-            when (command) {
-                is MessageCommand<*> -> messageCommands[match.id] = command
-                is SlashCommand<*, *> -> slashCommands[match.id] = command
-                is UserCommand<*> -> userCommands[match.id] = command
-            }
-        }
-
-        // Finally, we can remove anything that needs to be removed
-        toRemove.forEach {
-            logger.trace { "Removing ${it.type.name} command: ${it.name}" }
-            it.delete()
-        }
-
-        logger.info {
-            if (guild == null) {
-                "Finished synchronising global application commands"
-            } else {
-                "Finished synchronising application commands for guild ${guild.name}"
             }
         }
     }
 
-    // endregion
-
-    // region: Typed batch registration functions
-
-    /** Register multiple message commands. **/
-    public open suspend fun registerAll(vararg commands: MessageCommand<*>): List<MessageCommand<*>> =
-        commands.map {
-            try {
-                register(it) as MessageCommand<*>
-            } catch (e: KtorRequestException) {
-                logger.warn(e) {
-                    "Failed to register ${it.type.name} command: ${it.name}" +
-                        if (e.error?.message != null) {
-                            "\n        Discord error message: ${e.error?.message}"
-                        } else {
-                            ""
-                        }
-                }
-
-                null
-            } catch (t: Throwable) {
-                logger.warn(t) { "Failed to register ${it.type.name} command: ${it.name}" }
-
-                null
-            }
-        }.filterNotNull()
-
     /** Register multiple slash commands. **/
-    public open suspend fun registerAll(vararg commands: SlashCommand<*, *>): List<SlashCommand<*, *>> =
-        commands.map {
+    public open suspend fun <T : ApplicationCommand<*>> registerAll(vararg commands: T): List<T> =
+        commands.mapNotNull {
             try {
-                register(it) as SlashCommand<*, *>
-            } catch (e: KtorRequestException) {
-                logger.warn(e) {
-                    "Failed to register ${it.type.name} command: ${it.name}" +
-                        if (e.error?.message != null) {
-                            "\n        Discord error message: ${e.error?.message}"
-                        } else {
-                            ""
-                        }
+                when (it) {
+                    is SlashCommand<*, *> -> register(it) as T
+                    is MessageCommand<*> -> register(it) as T
+                    is UserCommand<*> -> register(it) as T
+                    else -> throw IllegalArgumentException(
```

Comment:
Don't make this exhaustive, we _want_ the compiler to tell us when all branches aren't accounted for.

Reply:
It needs to be exhaustive, it won't compile without it.

Reply:
Hmm, maybe I need to seal `ApplicationCommand`? Think that'd help?

Or, wait, then subtypes have to be in the same file, right? That's not great.

I guess if there's nothing we can do, we'll have to stick with it. But you can put a newline before the `else` branch. :>

Reply:
Fine by me